src/cicd/pipelines.ts

```diff
@@ -0,0 +1,321 @@
+import { Stack, StackProps, Stage } from 'aws-cdk-lib';
+import {
+  DetailType,
+  NotificationRule,
+} from 'aws-cdk-lib/aws-codestarnotifications';
+import { PolicyStatement } from 'aws-cdk-lib/aws-iam';
+import { Topic } from 'aws-cdk-lib/aws-sns';
+import {
+  CodeBuildStep,
+  CodePipeline,
+  CodePipelineSource,
+  IFileSetProducer,
+  ManualApprovalStep,
+  Step,
+} from 'aws-cdk-lib/pipelines';
+import { Construct, IConstruct } from 'constructs';
+import {
+  getBanditAction,
+  getCfnNagAction,
+  getCodeCommitSourceAction,
+  getSynthAction,
+  getTestsAction,
+} from './actions';
+import { toTitleCase } from './utils';
+
+export interface SourceActionProps {
+  readonly sourceAction?: CodePipelineSource;
+  readonly repositoryName: string;
+  readonly branch?: string;
+}
+
+export interface SynthActionProps {
+  readonly cdkVersion?: string;
+  readonly codeartifactRepository?: string;
+  readonly codeartifactDomain?: string;
+  readonly codeartifactDomainOwner?: string;
+  readonly rolePolicyStatements?: PolicyStatement[];
+  readonly synthAction?: CodeBuildStep;
+}
+
+export interface AddStageProps {
+  readonly stageId: string;
+  readonly stage: Stage;
+  readonly manualApprovals?: boolean;
+}
+
+export interface AddSecurityLintStageProps {
+  readonly stageName?: string;
+  readonly cloudAssemblyFileSet?: IFileSetProducer;
+}
+
+export interface AddTestStageProps {
+  readonly stageName?: string;
+  readonly cloudAssemblyFileSet?: IFileSetProducer;
+  readonly commands?: string[];
+}
+
+export interface AddNotificationsProps {
+  readonly notificationRule?: NotificationRule;
+}
+
+export interface AddCustomStageProps {
+  readonly stageName: string;
+  readonly steps: Step[];
+}
+
+export class CICDPipelineStack extends Stack {
+  readonly environmentId?: string;
+  readonly pipelineName?: string;
+  readonly pipelineId?: string;
+  public notificationRule?: NotificationRule;
+  public pipeline?: CodePipeline;
+  public pipelineKey?: IConstruct;
+  public sourceAction?: CodePipelineSource;
+  public synthAction?: CodeBuildStep;
+
+  constructor(
+    scope: Construct,
+    id: string,
+    environmentId: string,
+    pipelineName: string,
+    props?: StackProps,
+  ) {
+    super(scope, id, props);
+
+    this.environmentId = environmentId;
+    this.pipelineName = pipelineName;
+    this.pipelineId = id;
+  }
+
+  addSourceAction(props: SourceActionProps) {
+    var branch = props.branch ?? 'main';
+    this.sourceAction =
+      props.sourceAction ||
+      getCodeCommitSourceAction({
+        scope: this,
+        repositoryName: props.repositoryName,
+        branch: branch,
+      });
+    return this;
+  }
+
+  buildPipeline() {
+    /*
+    Build the pipeline structure.
+     Returns
+    -------
+    pipeline : CICDPipelineStack
+    CICDPipelineStack
+    */
+    this.pipeline = new CodePipeline(this, 'DDKCodePipeline', {
+      synth: this.synthAction!,
```

Comment:
It's bad practice to use the `!` operator. If we haven't define the `synthAction`, this is going to fail without giving us a good error message. I think a better way to process this would be:
```typescript
    if (this.synthAction === undefined) {
      throw new ArgumentError('Pipeline cannot be build without a synth action.');
    }
    this.pipeline = new CodePipeline(this, 'DDKCodePipeline', {
      synth: this.synthAction,
      crossAccountKeys: true,
      pipelineName: this.pipelineName,
      //cliVersion: Handle when Config() is decided on
    });
    return this;
  }
```

We could also add a test case for this scenario, to ensure the method throws an exception.

Reply:
Thanks for calling that out, I had forgotten to return to error handling and have this in a few places.